DREAM/Optimization/tsgParticleSwarm.cpp

```diff
@@ -0,0 +1,146 @@
+        }
+    }
+    initialized = true;
+}
+
+
+void ParticleSwarm(ObjectiveFunction f, int max_iterations, TasDREAM::DreamDomain inside, ParticleSwarmState &state,
+                   double inertia_weight, double cognitive_coeff, double social_coeff, std::function<double(void)> get_random01) {
+
+    // Only run the algorithm on properly initialized states.
+    if (!state.initialized) {
+        throw std::runtime_error("Particles have not been initialized in the input state object");
+    }
+
+    // Initialize helper variables and functions.
+    size_t num_dimensions = (size_t) state.getNumDimensions();
+    size_t num_particles = (size_t) state.getNumParticles();
+    std::vector<double> &particle_positions = state.getParticlePositionsRef();
+    std::vector<double> &particle_velocities = state.getParticleVelocitiesRef();
+    std::vector<double> &best_particle_positions = state.getBestParticlePositionsRef();
+    std::vector<double> &best_position = state.getBestPositionRef();
+    std::vector<double> prev_best_vals(num_particles), cur_best_vals(num_particles);
+    std::vector<bool> prev_is_inside(num_particles), cur_is_inside(num_particles);
+    double best_fval = std::numeric_limits<double>::max();
+    ObjectiveFunctionConstrained f_constrained = makeObjectiveFunctionConstrained(num_dimensions, f, inside);
+
+    // Create a lambda that updates the best particle and swarm positions based on function values and domain information.
+    auto update_best_positions = [&](const std::vector<double> prev_vals, const std::vector<double> cur_vals,
+                                     const std::vector<bool> prev_domain, const std::vector<bool> cur_domain)->void {
+        for (size_t i=0; i<num_particles; i++) {
+            bool is_smaller = (cur_vals[i] < prev_vals[i]) || !prev_domain[i];
+            if (cur_domain[i] && is_smaller) {
+                std::copy(particle_positions.begin() + i * num_dimensions, particle_positions.begin() + (i + 1) * num_dimensions,
+                          best_particle_positions.begin() + i * num_dimensions);
+                if (cur_vals[i] < best_fval) {
+                    std::copy(particle_positions.begin() + i * num_dimensions, particle_positions.begin() + (i + 1) * num_dimensions,
+                              best_position.begin());
+                    best_fval = cur_vals[i];
+                }
+            }
+        }
+    };
+
+    // Initialize the best particle and swarm function values, while updating the input state.
+    f_constrained(best_particle_positions, prev_best_vals, prev_is_inside);
+    f_constrained(particle_positions, cur_best_vals, cur_is_inside);
+    update_best_positions(prev_best_vals, cur_best_vals, prev_is_inside, cur_is_inside);
+    prev_best_vals = cur_best_vals;
+    prev_is_inside = cur_is_inside;
+
+    // Main algorithm starts here.
+    while(state.getNumIterations() < max_iterations) {
```

Comment:
You keep track of all iterations inside the `state` but `max_iterations` exits at the level of the algorithm. Let's take a minimalist approach here and count the iterations only for this current call to `ParticleSwarm()`.

The user can always count the iterations on their own and make judgement call on how to count which iteration. For example, initial iterations can be done with less social bias since we have very bad idea about the global minimum, later iterations can increase the bias as we begin to converge on the solution. Should the initial and later iterations count the same? Let the user figure it out.

Reply:
It seems like you are suggesting that we only have a local `iteration` variable in `ParticleSwarm()`, but how would the information be sent back to the user? Should I override `state.num_iterations`? Should I return `iteration`?

Reply:
ParticleSwarm doesn't have a convergence criteria, hence if the user asks for 10 iterations then we do 10 iterations.

When the user has a criteria (which would be another method), we can return the number of iterations either with `return num_iters` or as an output variable. If we make it part of the state (last resort), we can have something like `getNumLastIterations()`, but we don't keep some "global" history.

Reply:
Changed. I also removed `num_iterations` from the `ParticleSwarm` class definition.